assets/src/data/model/entity-factory/assertions.js

```diff
@@ -0,0 +1,283 @@
+import {
+	isShallowValidValueForField,
+	validateEnumType,
+	validateType,
+	validateTypeForField,
+} from './validators';
+import { maybeConvertFromValueObjectWithAssertions } from './extractors';
+
+/**
+ * Asserts whether the provided field value is a known value object.
+ *
+ * Note: this only asserts known value objects, if the value is not detected as
+ * a known value object it is passed back as is.
+ *
+ * @param {string} fieldName
+ * @param {*} fieldValue
+ * @param {Object} schema
+ * @throws InvalidDateTime
+ * @throws TypeError
+ */
+export const maybeAssertValueObject = ( fieldName, fieldValue, schema ) => {
+	if ( isDateTimeField( fieldName, schema ) ) {
+		DateTime.assertIsDateTime( fieldValue );
+	}
+	if ( isMoneyField( fieldName, schema ) ) {
+		Money.assertMoney( fieldValue );
+	}
+};
+
+/**
+ * Asserts whether the provided object is a valid model schema object.
+ *
+ * Currently, an object is considered a valid model schema if it has a
+ * 'properties' property.
+ *
+ * @param {*} schema
+ * @throws InvalidSchema
+ */
+export const assertValidSchema = ( schema ) => {
+	if ( ! isSchema( schema ) ) {
+		throw new InvalidSchema(
+			'This is an invalid schema for a model.',
+		);
+	}
+};
+
+/**
+ * Asserts that the given field exists in the provided schema and the shape for
+ * the schema entry on that field is expected.
+ *
+ * @param {string} modelName  The model the schema belongs to, this is used for
+ * error messages.
+ * @param {string} fieldName  The field being checked against the schema
+ * @param {Object} schema     The schema for the model used for validation
+ * @throws InvalidSchema
+ * @throws TypeError
+ */
+export const assertValidSchemaFieldProperties = (
+	modelName,
+	fieldName,
+	schema,
+) => {
+	if ( isUndefined( schema[ fieldName ] ) ) {
+		throw new TypeError(
+			sprintf(
+				'The given "%s" fieldName does not have a defined schema ' +
+				'for the model "%s"',
+				fieldName,
+				modelName,
+			),
+		);
+	}
+	if ( schema[ fieldName ].type === 'object' ) {
+		if ( isUndefined( schema[ fieldName ].properties ) ) {
+			throw new InvalidSchema(
+				sprintf(
+					'The schema for the field %s on the model %s is of type ' +
+					'"object" but does not have a properties property.',
+					fieldName,
+					modelName
+				)
+			);
+		}
+		if ( isUndefined( schema[ fieldName ].properties.raw ) ) {
+			throw new InvalidSchema(
+				sprintf(
+					'The schema for the field %s on the model %s is of type ' +
+					'"object" but does not have a raw property in it\'s ' +
+					'"properties" property.',
+					fieldName,
+					modelName
+				)
+			);
+		}
+		if ( isUndefined( schema[ fieldName ].properties.raw.type ) ) {
+			throw new InvalidSchema(
+				sprintf(
+					'The schema for the field %s on the model %s is of type ' +
+					'"object" and has a properties.raw property, however there' +
+					'is no "type" defined for the raw property.',
+					fieldName,
+					modelName
+				),
+			);
+		}
+	}
+};
```

Comment:
When there's an Error thrown for a function that could get called in UI context I translate the strings.  But generally for Error's I don't bother localizing (especially all these ones) because it's only intended to show up for developers (these errors get thrown on the construction of an entity which typically happens in low-level code).